go-packages/meep-model/model.go

```diff
@@ -943,6 +957,737 @@ func (m *Model) GetNodeContext(name string) (ctx interface{}) {
+//              Returned value is of type []interface{}
+func (m *Model) GetNetworkLocationNodesByFilter(filter *NodeFindFilter) (node []interface{}) {
+	return m.GetNodesByFilter("networkLocation", filter)
+}
+
+// GetPhysicalLocationNodesByFilter - Get list of interfaces that match the filter providef for physical locations
+//              Returned value is of type []interface{}
+func (m *Model) GetPhysicalLocationNodesByFilter(filter *NodeFindFilter) (node []interface{}) {
+	return m.GetNodesByFilter("physicalLocation", filter)
+
+}
+
+// GetProcessNodesByFilter - Get list of interfaces that match the filter provided for processes
+//              Returned value is of type []interface{}
+func (m *Model) GetProcessNodesByFilter(filter *NodeFindFilter) (node []interface{}) {
+	return m.GetNodesByFilter("process", filter)
+}
+
+// GetNodesByChildLevelFilter - Get list of interfaces that match the filter provided for children of provided hierachy level
+//              Returned value is of type []interface{}
+func (m *Model) GetNodesByChildLevelFilter(queriedLevel string, filter *NodeFindFilter, nList []*Node) (node []interface{}) {
+
+	switch queriedLevel {
+	case "domain":
+		return m.GetDomainNodesByChildLevelFilter(filter, nList)
+	case "zone":
+		return m.GetZoneNodesByChildLevelFilter(filter, nList)
+	case "networkLocation":
+		return m.GetNetworkLocationNodesByChildLevelFilter(filter, nList)
+	case "physicalLocation":
+		return m.GetPhysicalLocationNodesByChildLevelFilter(filter, nList)
+	case "process":
+		return m.GetProcessNodesByChildLevelFilter(filter, nList)
+	default:
+	}
+
+	return nil
+}
+
+// GetProcessNodesByChildLevelFilter - Get list of interfaces that match the filter provided for children of processes
+//              Returned value is of type []interface{}
+func (m *Model) GetProcessNodesByChildLevelFilter(filter *NodeFindFilter, nList []*Node) (node []interface{}) {
+	//no children to process so no check necessary
+	//just converting nList to interface
+	var nListInterface []interface{}
+	for _, currentNode := range nList {
+		process := currentNode.object.(*dataModel.Process)
+		nListInterface = append(nListInterface, process)
+	}
+	return nListInterface
+}
+
+// GetDomainNodesByChildLevelFilter - Get list of interfaces that match the filter provided for children of domains
+//              Returned value is of type []interface{}
+func (m *Model) GetDomainNodesByChildLevelFilter(filter *NodeFindFilter, nList []*Node) (node []interface{}) {
+
+	var foundDomain, validDomainChildren bool
+	var foundZone, validZoneChildren bool
+	var foundNl, validNlChildren bool
+	var foundPl, validPlChildren bool
+
+	var nListInterface []interface{}
+
+	for _, currentNode := range nList {
+		foundDomain = true
+		domainObj := currentNode.object.(*dataModel.Domain)
+
+		//create a deepCopy
+		var domain dataModel.Domain
+		byt, _ := json.Marshal(domainObj)
+		json.Unmarshal(byt, &domain)
+
```

Comment:
You are still parsing the scenario here, which we are trying to avoid.
The filter function should only check the required filter criteria. For example:
```
func filter(node) bool {
  if filter.DomainName != "" {
    filterNodeByZone(node)
  }
  if filter.ZoneName != "" {
    valid := filterNodeByZone(node, filter.ZoneName)
    if !valid {
      return false
    } 
  }
  // etc.
 return true
}

func filterNodeByZone(node, zoneName) bool {
  If node type is above zone {
    // Get nodes with type Zone from nodemap
    // Loop through them and see if they have the current node as a parent in their context
  } else {
    // Check the current node context to see if the zone name is there
  }
}

//etc.
```

Reply:
no more parsing